internal/agent/hooks/gruboptions.go

```diff
@@ -44,22 +94,112 @@ func (b GrubFirstBootOptions) Run(c config.Config, _ v1.Spec) error {
 }
 
 // grubOptions sets the grub options in the grubenv file
-// It mounts the OEM partition if not already mounted
-// If its mounted but RO, it remounts it as RW
-func grubOptions(c config.Config, opts map[string]string) error {
-	_ = machine.Umount(cnst.OEMDir)
-	_ = machine.Umount(cnst.OEMPath)
-
-	c.Logger.Logger.Debug().Msg("Mounting OEM partition")
-	_ = machine.Mount(cnst.OEMLabel, cnst.OEMPath)
+// It ALWAYS writes to STATE partition (grubenv) as that's what GRUB reads
+// It optionally writes to OEM partition (grub_oem_env) only if OEM is not encrypted
+func grubOptions(c config.Config, opts map[string]string, oemEncrypted bool) error {
+	var firstErr error
+
+	// Always write to STATE partition (grubenv) - this is what GRUB reads during boot
+	_ = machine.Umount(cnst.StateDir)
+	c.Logger.Logger.Debug().Msg("Mounting STATE partition")
+	_ = machine.Mount(cnst.StateLabel, cnst.StateDir)
 	defer func() {
-		c.Logger.Logger.Debug().Msg("Unmounting OEM partition")
-		_ = machine.Umount(cnst.OEMPath)
+		c.Logger.Logger.Debug().Msg("Unmounting STATE partition")
+		_ = machine.Umount(cnst.StateDir)
 	}()
 
-	err := utils.SetPersistentVariables(filepath.Join(cnst.OEMPath, "grubenv"), opts, &c)
+	err := utils.SetPersistentVariables(filepath.Join(cnst.StateDir, cnst.GrubEnv), opts, &c)
 	if err != nil {
-		c.Logger.Logger.Error().Err(err).Str("grubfile", filepath.Join(cnst.OEMPath, "grubenv")).Msg("Failed to set grub options")
+		c.Logger.Logger.Error().Err(err).Str("grubfile", filepath.Join(cnst.StateDir, cnst.GrubEnv)).Msg("Failed to set grub options in STATE")
+		firstErr = err
+	} else {
+		c.Logger.Logger.Info().Str("grubfile", filepath.Join(cnst.StateDir, cnst.GrubEnv)).Msg("Successfully set grub options in STATE")
 	}
-	return err
+
+	// Only write to OEM if it's not encrypted (GRUB can't read it if encrypted anyway)
```

Comment:
mmmh, does this means that if we encrypt oem, then boot assesment wont work, nor the bootentries command ? They all depend on config files available in there in order to work IIRC....

Reply:
yep, they rely in opening the config files under /oem to check for and write values so indeed they would never work under encryption :(